File: files/100-199/137/CH-137.R
library(tidyverse)
library(readxl)
library(janitor)

path = "files/CH-137 Table Transformation.xlsx"
input = read_excel(path, range = "C2:D17")
test  = read_excel(path, range = "F2:H12") %>% mutate(Date = as.Date(Date))

result = input %>%
  mutate(Date = ifelse(is.na(`Column 2`), as.numeric(`Column 1`), NA)) %>%
  fill(Date) %>%
  na.omit() %>%
  transmute(Date = excel_numeric_to_date(Date), Product = `Column 1`, Quantity = `Column 2`)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE

File: files/100-199/137/CH-137.py
import pandas as pd

path = "CH-137 Table Transformation.xlsx"

input = pd.read_excel(path, usecols="C:D", skiprows=1, nrows=15)
test = pd.read_excel(path, usecols="F:H", skiprows=1, nrows=10, parse_dates=['Date'], dtype={'Quantity': 'float64'})

input['Date'] = input.apply(lambda row: row['Column 1'] if pd.isna(row['Column 2']) else None, axis=1)
input['Date'] = input['Date'].ffill()
input = input.dropna()

result = input[['Date', 'Column 1', 'Column 2']].rename(columns={'Column 1': 'Product', 'Column 2': 'Quantity'}).reset_index(drop=True)

print(test.equals(result)) # True
